File: dart/04-Stack-Queue/LoopQueue.dart
import 'Queue.dart';

class LoopQueue<T> implements Queue<T> {
  List? arr;
  int front = 0, tail = 0;
  int size = 0;

  LoopQueue() {
    arr = List.filled(10, false);
  }

  // LoopQueue.capcity(int capacity){
  //   arr = List.filled(capacity+1, false);
  // }

  int getCapacity() {
    //目前这种方式浪费一个空间
    int capacity = arr!.length - 1;
    return capacity;
  }

  @override
  dequeue() {
    if (isEmpty()) {
      throw Exception("Cannot dequeue from an empty queue");
    }
    T ret = arr![front];
    arr![front] = null;
    front = (front + 1) % arr!.length;
    size--;
    if (size == getCapacity() / 4 && getCapacity() / 2 != 0) {
      resize((getCapacity() / 2) as int);
    }
    return ret;
  }

  void resize(int newCapacity) {
    List? newArr = List.filled(newCapacity + 1, false);
    for (int i = 0; i < arr!.length; i++) {
      newArr[i] = arr![(i + front) % arr!.length];
    }
    arr = newArr;
    front = 0;
    tail = size;
  }

  @override
  void enqueue(e) {
    if ((tail + 1) % arr!.length == front) resize(getCapacity() * 2);
    arr![tail] = e;
    tail = (tail + 1) % arr!.length;
    size++;
  }

  @override
  getFront() {
    if (isEmpty()) {
      throw Exception("Queue is empty.");
    }
    return arr![front];
  }

  @override
  int getSize() {
    return arr!.length;
  }

  @override
  bool isEmpty() {
    return front == tail;
  }

  @override
  String toString() {
    StringBuffer buffer = new StringBuffer();
    buffer.write("Queue: size = $size , capacity = ${getCapacity()} \n");
    buffer.write('front [');
    for (int i = front; i != tail; i = (i + 1) % arr!.length) {
      buffer.write(arr![i]);
      if ((i + 1) % arr!.length != tail) buffer.write(", ");
    }
    buffer.write("] tail");
    return buffer.toString();
  }
}

File: python/loopqueue.py
'''
    循环队列
'''
class LoopQueue(object):

    def __init__(self, n=10):
        self.arr = [None] * (n)  
        self.front = 0
        self.tail = 0
        self.size = 0

    def getCapacity(self):
        capacity = len(self.arr) - 1;
        return capacity;

    def dequeue(self):
        if self.isEmpty() :
            raise Exception("Cannot dequeue from an empty queue")
        ret = self.arr[self.front]
        self.arr[self.front] = None
        self.front = (self.front + 1)%len(self.arr)
        self.size -=1
        if((self.size == self.getCapacity()//4)) and (self.getCapacity()/2 !=0):
            resize(self.getCapacity()//2)
        return ret

    def resize(self,newCapacity):
        newArr = [None]*(newCapacity+1)
        for i in range(0,len(self.arr)):
            newArr[i] = self.arr[(i+self.front)%len(self.arr)];
        self.arr = newArr
        self.front = 0;
        self.tail = self.size;
    
    def enqueue(self,element):
        print('当前大小:',self.size)
        if ((self.tail+1)%len(self.arr)) == self.front:
            self.resize(self.getCapacity()*2)
        self.arr[self.tail] = element
        self.tail = (self.tail +1)% len(self.arr)
        self.size += 1
        print('当前大小:',self.size)

    def getFront(self):
        if(self.isEmpty()):
            raise Exception("Queue is empty.")        
        return self.arr[self.front]

    def getSize(self):
        return len(self.arr)
    
    def isEmpty(self):
        return self.front == self.tail
        
    def __str__(self):
        return "front ["+" ".join(map(str, self.arr)) + "] tail"

if __name__ == '__main__':
    loopQueue = LoopQueue();
    for i in range(0,30):
        loopQueue.enqueue(i);
        print(loopQueue)

      
      